core/src/km_core_state_context_set_if_needed.cpp

```diff
@@ -26,115 +26,226 @@
+  }
 
-  if(set_context_from_string(cached_context, new_cached_context) != KM_CORE_STATUS_OK) {
-    return do_fail(app_context, cached_context, "could not set new cached context");
+  if (!replace_context(context_changed, cached_context, new_cached_context)) {
+    switch (context_changed.type) {
+    case CONTEXT_DIFFERENT:
+      return do_fail(app_context, cached_context, "could not set new cached context");
+    case CONTEXT_SHORTER:
+      return do_fail(app_context, cached_context, "could not prepend new cached context");
+    case CONTEXT_LONGER:
+      return do_fail(app_context, cached_context, "could not shrink cached context");
+    }
   }
 
   return KM_CORE_CONTEXT_STATUS_UPDATED;
 }
 
+bool
+replace_context(
+  context_change_type context_changed,
+  km_core_context *context,
+  km_core_cp const *new_context
+) {
+  if (context_changed.type == CONTEXT_DIFFERENT) {
+    if (set_context_from_string(context, new_context) != KM_CORE_STATUS_OK) {
+      return false;
+    }
+  } else if (context_changed.type == CONTEXT_SHORTER) {
+    km_core_context_item *new_context_items;
+    if (context_items_from_utf16(new_context, &new_context_items) != KM_CORE_STATUS_OK ||
+        context_prepend(context, new_context_items, context_changed.end_difference) != KM_CORE_STATUS_OK) {
+      km_core_context_items_dispose(new_context_items);
+      return false;
+    }
+  } else {
+    assert(context_changed.type == CONTEXT_LONGER);
+    if (context_shrink(context, context_changed.end_difference, false) != KM_CORE_STATUS_OK) {
+      return false;
+    }
+  }
+  return true;
+}
+
 /**
  * Returns true if the current keyboard processor wants a normalized cached context
  */
 bool should_normalize(km_core_state *state) {
   return state->processor().supports_normalization();
 }
 
+/**
+ * Returns the previous context item that is a character skipping any
+ * markers, or NULL if there are no more characters before to
+ * `context_items`.
+ */
+km_core_context_item const *
+context_previous_char(
+  km_core_context_item const *context_items,
+  km_core_context_item const *context_start
+) {
+  if (context_items == context_start) {
+    return NULL;
+  }
+  while (context_items > context_start) {
+    context_items--;
+    if (context_items->type == KM_CORE_CT_CHAR) {
+      return context_items;
+    }
+    // skip over marker
+  }
+  return NULL;
+}
+
 /**
  * Returns true if the internal app context does not need to be updated to the new
  * app context
- *
- * TODO: #10100 will alter some of the assumptions here
  */
-bool is_context_valid(km_core_cp const * new_app_context, km_core_cp const * app_context) {
-  if (new_app_context == nullptr || app_context == nullptr || *app_context == '\0') {
+context_change_type
+is_context_unchanged(
+  km_core_cp const *new_context_string,
+  km_core_context *context
+) {
+  context_change_type change_type({CONTEXT_DIFFERENT, 0});
+  if (new_context_string == nullptr || context == nullptr) {
```

Comment:
Suggest we add assert(new_context_string != nullptr) and assert(context != nullptr) for debug builds